src/nrniv/neuronapi.cpp

```diff
@@ -0,0 +1,519 @@
+#include "neuronapi.h"
+
+#include "../../nrnconf.h"
+#include "hocdec.h"
+#include "nrniv_mf.h"
+#include "nrnmpi.h"
+#include "nrnmpiuse.h"
+#include "ocfunc.h"
+#include "ocjump.h"
+#include "parse.hpp"
+#include "section.h"
+
+/// A public face of hoc_Item
+struct nrn_Item: public hoc_Item {};
+
+struct SectionListIterator {
+    SectionListIterator(nrn_Item*);
+    Section* next(void);
+    int done(void);
+
+  private:
+    hoc_Item* initial;
+    hoc_Item* current;
+};
+
+struct SymbolTableIterator {
+    SymbolTableIterator(Symlist*);
+    char const* next(void);
+    int done(void);
+
+  private:
+    Symbol* current;
+};
+
+/****************************************
+ * Connections to the rest of NEURON
+ ****************************************/
+extern int nrn_nobanner_;
+extern int diam_changed;
+extern int nrn_try_catch_nest_depth;
+extern "C" void nrnpy_set_pr_etal(int (*cbpr_stdoe)(int, char*), int (*cbpass)());
+int ivocmain_session(int, const char**, const char**, int start_session);
+void simpleconnectsection();
+extern Object* hoc_newobj1(Symbol*, int);
+extern void nrn_change_nseg(Section*, int);
+extern std::tuple<int, char**> nrn_mpi_setup(int argc, char** argv);
+
+extern "C" {
+
+/****************************************
+ * Initialization
+ ****************************************/
+
+int nrn_init(int argc, const char** argv) {
+    nrn_nobanner_ = 1;
+    auto [final_argc, final_argv] = nrn_mpi_setup(argc, const_cast<char**>(argv));
+    errno = 0;
+    return ivocmain_session(final_argc, (const char**) final_argv, nullptr, 0);
+}
+
+void nrn_stdout_redirect(int (*myprint)(int, char*)) {
+    // the first argument of myprint is an integer indicating the output stream
+    // if the int is 1, then stdout, else stderr
+    // the char* is the message to display
+    nrnpy_set_pr_etal(myprint, nullptr);
+}
+
+/****************************************
+ * Sections
+ ****************************************/
+
+Section* nrn_section_new(char const* const name) {
+    // TODO: check for memory leaks; should we free the symbol, pitm, etc?
```

Comment:
just a question: still relevant? same for the other `TODO` comments. Fine to have those but would be nice to confirm all those are still relevant.

Reply:
If there's something I'm not very happy with is this. For me it's hard to understand the full path of code allocations, and why the memory sanitizer didn't complain about this one. But I don't know if it's safe to deallocate either. it's a matter of time investment.